core/src/wsv.rs

```diff
@@ -1080,37 +1109,33 @@ impl WorldStateView {
         self.events_buffer.borrow_mut().push(event.into());
     }
 
-    /// The same as [`Self::modify_validator_multiple_events`] except closure `f` returns a single [`PermissionValidatorEvent`].
+    /// Get constant view to a *Runtime Permission Validator*.
     ///
-    /// # Errors
-    /// Forward errors from [`Self::modify_validators_multiple_events`]
-    pub fn modify_validators<F>(&self, f: F) -> Result<(), Error>
-    where
-        F: FnOnce(&crate::validator::Chain) -> Result<PermissionValidatorEvent, Error>,
-    {
-        self.modify_validators_multiple_events(move |chain| f(chain).map(std::iter::once))
-    }
-
-    /// Get [`crate::validator::Chain`] and pass it to `closure` to modify it
+    /// Performs lazy upgrade of the validator if [`Upgrade`] instruction was executed.
     ///
-    /// # Errors
-    /// Forward errors from `f`
-    pub fn modify_validators_multiple_events<I, F>(&self, f: F) -> Result<(), Error>
-    where
-        I: IntoIterator<Item = PermissionValidatorEvent>,
-        F: FnOnce(&crate::validator::Chain) -> Result<I, Error>,
-    {
-        self.modify_world_multiple_events(|world| {
-            f(&world.validators)
-                .map(|events| events.into_iter().map(WorldEvent::PermissionValidator))
-        })
-    }
-
-    /// Get constant view to the chain of validators.
+    /// # Panic
     ///
-    /// View guarantees that no interior-mutability can be performed.
-    pub fn validators_view(&self) -> crate::validator::ChainView {
-        self.world.validators.view()
+    /// Panics if validator is not initialized.
+    /// Possible only before applying genesis.
+    pub fn validator_view(&self) -> crate::validator::View {
+        {
+            let mut upgraded_validator_write = self.world.upgraded_validator.write();
+            if let Some(upgraded_validator) = upgraded_validator_write.take() {
+                let mut validator_write = self.world.validator.write();
+                validator_write.replace(upgraded_validator);
+            }
+        }
+
+        #[cfg(test)]
+        {
+            let mut validator_write = self.world.validator.write();
+            if validator_write.is_none() {
+                let _ = validator_write.insert(Validator::default());
```

Comment:
placeholder is special and in this situation `let _ = validator_write.insert(Validator::default());` should be exactly the same as `validator_write.insert(Validator::default());`. Maybe you want to use something like `let _validator = ...`

Reply:
`insert()` here returns `&mut Validator`, yes. But what the difference here between `_` and `_validator`?

Reply:
`clippy` warns about unused result. If there will be much difference between `_` and `_validator`, then I think `clippy` would warn me too